3_time_series/Statistical_analysis.md

```diff
@@ -0,0 +1,749 @@
+    predictions =  []
+    prediction_range = pd.date_range(split,  dataframe.index[-1], freq='MS')
+    for i, cur_date in enumerate(prediction_range):
+        if(verbose): print(i, " iteration")
+        train_data = dataframe[:cur_date].reset_index(drop=True)
+        train_data["lag_1"] = train_data["Num_Passengers"].shift(1)
+        train_data = train_data[1:]
+        
+        train_data["diff"] = train_data["Num_Passengers"].diff(1)
+        train_data = train_data[1:]
+        
+        train_data["lag_diff_12"] = train_data["diff"].shift(12)
+        train_data = train_data[12:]
+       
+        train_col = ["lag_1", "lag_diff_12"]
+        fit, pred = linear_regression(train_data[train_col].iloc[:-1, :], train_data["Num_Passengers"].iloc[:-1], 
+                                     train_data[train_col].iloc[-1:, :])
+        model_fit.append(fit)
+        predictions.append(pred)
+        
+    predictions_df = pd.DataFrame({'Month': prediction_range,
+                   'Num_Passengers': predictions})
+    predictions_df = predictions_df.set_index('Month')
+    
+    return model_fit, predictions_df
+```
+
+```python
+def fit_model_arima(order, dataframe, split, method=None, verbose=False):
+    model_fit = []
+    predictions =  []
+    
+    for i, cur_date in enumerate(pd.date_range(split,  dataframe.index[-1], freq='MS')):
+        if(verbose): print(i, " iteration")
+        model = ARIMA(dataframe[:cur_date - timedelta(days=1)], order = order)
+        fit = model.fit(method=method)
+        pred = fit.predict(start=cur_date, end=cur_date) 
+        
+        model_fit.append(fit)
+        predictions.append(pred)
+    
+    predictions_df = pd.DataFrame(pd.concat(predictions, axis=0), columns=['Num_Passengers'])
+    return model_fit, predictions_df
+```
+
+```python
+def fit_model_sarimax(order, seasonal_order, dataframe, split, verbose=False):
+    model_fit = []
+    predictions =  []
+    
+    for i, cur_date in enumerate(pd.date_range(split,  dataframe.index[-1], freq='MS')):
+        if(verbose): print(i, " iteration")
+        model = SARIMAX(dataframe[:cur_date - timedelta(days=1)], order=order, seasonal_order=seasonal_order)
+        fit = model.fit()
+        pred = fit.predict(start=cur_date, end=cur_date) 
+        
+        model_fit.append(fit)
+        predictions.append(pred)
+    
+    predictions_df = pd.DataFrame(pd.concat(predictions, axis=0), columns=['Num_Passengers'])
+    return model_fit, predictions_df
+```
+
+```python
+def get_split_date(data, split_coef):
+    if split_coef<0 or split_coef>1:
+        raise ValueError("Split coefficient should be in range [0,1]")
+    split = math.floor(len(data) * split_coef)
+    split = split if split<len(data) else split-1
+    return data.index[split]
+```
+
+```python
+def split_train_test(df_to_split, train_percents):
+    split_date = get_split_date(df_to_split, train_percents)
+
+    train_data = df_to_split[:split_date - timedelta(days=1)]
```

Comment:
do not need timedelta, by design `[:3]` and `[3:]` are non overlapping intervals. meaning `(-inf, 3)` and `<3, inf)` in math notation

Reply:
It's what I expected but looks like it's working differently with time index - [Time series](https://pandas.pydata.org/docs/user_guide/timeseries.html#slice-vs-exact-match:~:text=This%20type%20of%20slicing%20will%20work%20on%20a%20DataFrame%20with%20a%20DatetimeIndex%20as%20well.%20Since%20the%20partial%20string%20selection%20is%20a%20form%20of%20label%20slicing%2C%20the%20endpoints%20will%20be%20included.%20This%20would%20include%20matching%20times%20on%20an%20included%20date%3A)